server/src/main/java/org/elasticsearch/rest/RestHandler.java

```diff
@@ -59,4 +64,101 @@ default boolean supportsContentStream() {
     default boolean allowsUnsafeBuffers() {
         return false;
     }
+
+    /**
+     * The list of {@link Route}s that this RestHandler is responsible for handling.
+     */
+    default List<Route> routes() {
+        return Collections.emptyList();
+    }
+
+    /**
+     * A list of routes handled by this RestHandler that are deprecated and do not have a direct
+     * replacement. If changing the {@code path} or {@code method} of a route,
+     * use {@link #replacedRoutes()}.
+     */
+    default List<DeprecatedRoute> deprecatedRoutes() {
+        return Collections.emptyList();
+    }
+
+    /**
+     * A list of routes handled by this RestHandler that have had their {@code path} and/or
+     * {@code method} changed. The pre-existing {@code route} will be registered
+     * as deprecated alongside the updated {@code route}.
+     */
+    default List<ReplacedRoute> replacedRoutes() {
+        return Collections.emptyList();
+    }
+
+    class Route {
+
+        private final String path;
+        private final Method method;
+
+        public Route(Method method, String path) {
+            this.path = path;
+            this.method = method;
+        }
+
+        public String getPath() {
+            return path;
+        }
+
+        public Method getMethod() {
+            return method;
+        }
+    }
+
+    /**
+     * Represents an API that has been deprecated and is slated for removal.
+     */
+    class DeprecatedRoute extends Route {
+
+        private final String deprecationMessage;
+        private final DeprecationLogger logger;
+
+        public DeprecatedRoute(Method method, String path, String deprecationMessage, DeprecationLogger logger) {
```

Comment:
In a follow-up, let's get these loggers out of here. It doesn't make sense they are put of the route, and reflecting on it, I don't think we need a dedicated deprecation logger for every REST handler that deprecates some method. I think we can create a top-level RestController deprecation logger for this purpose.